File: cpp/91. [M] Decode Ways.cpp
// https://leetcode.com/problems/container-with-most-water/

#include <vector>  
#include <iostream>
#include <string>

using namespace std;


class Solution {
public:
    int numDecodings(string s) {
        int len = s.size();
        if (len == 0 || s.at(0) == '0')
            return 0;
        int prev2 = 1, prev1 = 1, prevDigit = s.at(0) - '0';

        for (int i = 1; i < len; i++) {
            int currDigit = s.at(i) - '0';
            if (currDigit == 0 && (prevDigit > 2 || prevDigit == 0))
                return 0;
            processDigit(prev2, prev1, prevDigit, currDigit);
        }
        return prev1;
    }

    void processDigit(int &prev2, int &prev1, int &prevDigit, int currDigit) {
        int curr = prev1;
        if (currDigit == 0) {
            curr = prev2;
        }
        // cout << prevDigit << ' ' << currDigit << endl;
        else if (prevDigit == 1 || (prevDigit == 2 && currDigit <= 6)) {
            curr += prev2;
        }
        prev2 = prev1;
        prev1 = curr;
        prevDigit = currDigit;
    }
};

int main() {
    Solution sol = Solution();
    cout << sol.numDecodings("102") << endl;
}

File: python/91. [M] Decode Ways.py
class Solution:
    def num_decodings(self, s: str) -> int:
        self.d = [-1] * len(s)
        return int(self.num_decodings_from_n(s, 0))

    def num_decodings_from_n(self, s: str, n: int) -> int:
        if len(s) == n:
            return 1
        if len(s) == n - 1:
            return 0
        if self.d[n] != -1:
            return self.d[n]
        if s[n] == '0':
            self.d[n] = 0
            return 0

        one_digit = self.num_decodings_from_n(s, n + 1)
        two_digits = 0 if int(s[n:n+2]) > 26 else self.num_decodings_from_n(s, n + 2)
        self.d[n] = one_digit + two_digits
        return self.d[n]
